### ISSUE TITLE
trial --testmodule should be usable without causing test methods to run repeatedly

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#11129]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#1563|
|Type|defect|
|Created|2006-03-13 03:10:19Z|

For example, ```trial --testmodule=foo --testmodule=bar``` should only run the contents of test_foobar once, even if foo.py and bar.py both declare that their test-case-name is test_foobar.

<details><summary>Searchable metadata</summary>

```
trac-id__1563 1563
type__defect defect
reporter__exarkun exarkun
priority__highest highest
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__trial trial
keywords__ 
time__1142219419000000 1142219419000000
changetime__1152423155000000 1152423155000000
version__ 
owner__jml jml
cc__jml,glyph,exarkun
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @jml|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun commented|
|-|-|

Here's another example of --testmodule misbehavior:

```
#!sh
exarkun@kunai:~$ cat foo.py
# -*- test-case-name: foo -*-
from twisted.trial import unittest

class Foo(unittest.TestCase):
    def setUpClass(self):
        print 'setUpClass'
    
    def tearDownClass(self):
        print 'tearDownClass'


    def testFoo(self):
        pass
exarkun@kunai:~$ trial --testmodule foo.py
Running 2 tests.
foo
  Foo
    testFoo ... setUpClass
                                                           [OK]
    testFoo ... tearDownClass
                                                           [OK]

-------------------------------------------------------------------------------
Ran 2 tests in 0.060s

PASSED (successes=2)
exarkun@kunai:~$ 
```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun commented|
|-|-|

Here is another problem/example of this problem:
```
exarkun@kunai:~$ cat barthing
# -*- test-case-name: zang.test.test_foo -*-
exarkun@kunai:~$ cat zang/test/test_foo.py

from twisted.trial import unittest

class Foo(unittest.TestCase):
    def setUpClass(self):
        print 'setUpClass'
    
    def tearDownClass(self):
        print 'tearDownClass'


    def testFoo(self):
        pass


class Bar(Foo):
    def testBar(self):
        pass
exarkun@kunai:~$ trial --testmodule barthing
Running 6 tests.
zang.test.test_foo
  Bar
    testBar ... setUpClass
                                                           [OK]
    testFoo ...                                                            [OK]
  Foo
    testFoo ... setUpClass
                                                           [OK]
  Bar
    testBar ...                                                            [OK]
    testFoo ... tearDownClass
                                                           [OK]
  Foo
    testFoo ... tearDownClass
                                                           [OK]

-------------------------------------------------------------------------------
Ran 6 tests in 0.064s

PASSED (successes=6)
exarkun@kunai:~$ trial --testmodule zang/test/test_foo.py
Running 3 tests.
zang.test.test_foo
  Bar
    testBar ... setUpClass
                                                           [OK]
    testFoo ... tearDownClass
                                                           [OK]
  Foo
    testFoo ... setUpClass
tearDownClass
                                                           [OK]
-------------------------------------------------------------------------------
Ran 3 tests in 0.059s

PASSED (successes=3)
```

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph set owner to @glyph<br>@glyph set status to `assigned`|
|-|-|

OK, this does not look hard to fix at all, and twisted-dev.el is totally broken now.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph set owner to @jerub<br>@glyph set status to `new`|
|-|-|

Please review!

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @glyph|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @glyph|
|-|-|

Branch looks good, for what it does.  Go ahead and merge, but leave the ticket open so someone can take care of the rest of the issue.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph commented|
|-|-|

(In [16607]) Fix trial enough to work with recent twisted-dev.el changes.

Author: glyph

Reviewer: exarkun

Refs [#11129](https://github.com/twisted/twisted/issues/11129)

Previously, --testmodule would add every module to the test run twice, causing
all tests to be invoked twice.  Sadly, many tests are not safe to invoke twice
in the same process, and would fail spuriously on their second run.

This change fixes that, but according to exarkun does not address all the
issues described in [#11129](https://github.com/twisted/twisted/issues/11129).

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph set owner to @jml|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)|@jml set owner to @glyph|
|-|-|

I've fixed the remaining issues (I think -- the comments don't exactly make it clear) in [branches/nail-1563](branches/nail-1563)

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)|@jml set owner to @jerub|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="jerub's avatar" src="https://avatars.githubusercontent.com/u/315461?s=50" width="50" height="50">](https://github.com/jerub)|@jerub set owner to @jml|
|-|-|

looks good, please merge.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)|@jml set status to `closed`|
|-|-|

(In [16819]) Fix the remaining issues with --testmodule

 * Fixes [#11129](https://github.com/twisted/twisted/issues/11129)
 * Reviewer: jerub
 * Author: jml

Queues all of the tests to run in a set, which means that we won't ever
be able to run duplicate tests from the command line.  That seems to be 
the remaining problem reported in [#11129](https://github.com/twisted/twisted/issues/11129).

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set status to `reopened`|
|-|-|

This doesn't seem to be fixed.  Consider the behavior of

```
trial --testmodule=twisted/protocols/policies.py --testmodule=twisted/test/test_policies.py
```

in trunk@16862

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)|@jml commented|
|-|-|

Thanks.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)|@jml set owner to @exarkun<br>@jml set status to `new`|
|-|-|

Fix in [branches/another-facet-1563](branches/another-facet-1563)

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)|@jml set owner to @jml|
|-|-|

Ugh.  For some reason, _nothing_ works in this branch :(

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)|@jml set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)|@jml set owner to @exarkun<br>@jml set status to `new`|
|-|-|

OK. I _think_ this is done, but given the embarrassing history of this ticket, I'll say nothing stronger than that.

Ready for review.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)|@jml commented|
|-|-|

Merged forward to [branches/another-facet-1563-2](branches/another-facet-1563-2)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @jml<br>@exarkun set status to `new`|
|-|-|

Is the removal of suppresswarnings in this branch intentional?  I don't really object (since the feature was already gone and the change only amounts to a handful of deleted lines) but make sure you mention this change when you merge the branch, if it is really intentional.

Fixes look good, interactive testing shows it behaves how I want it to behave.  The only change I would suggest is to add docstrings to all of the new and changed test methods and for loadByNames.

Once added, no need to re-assign to me for review, just merge.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)|@jml set status to `closed`|
|-|-|

(In [17511]) Really make sure --testmodule doesn't let duplicates in. Also remove
deprecated --suppresswarnings option.

 * Author: jml
 * Reviewer: exarkun
 * Fixes [#11129](https://github.com/twisted/twisted/issues/11129)

Takes the code that turns a bunch of names into test cases and moves 
it into one place: TestLoader.loadTestNames. The method turns the 
strings into Python objects, then filters out duplicates, then 
returns a suite of all the tests.
